### PR TITLE
Alanwaketan/rebase

### PR DESCRIPTION
**Pull request by alanwaketan**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by alanwaketan** [comment]
@jonb377 This is the rebase I mentioned. Once you approve, and I will force push this to llama2-google-next-training.

File: SPMD_USER_GUIDE.md [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+```
+
+Once the QueuedResource’s status becomes ACTIVE, the TPU VMs are provisioned and ready to run your training workload. See the [official guide](https://cloud.google.com/tpu/docs/queued-resources) for more information on a QueuedResource’s life cycle.
+
+
+### HF Llama 2 Environment Setup
+
+Here PyTorch 2.1 and PyTorch/XLA 2.1 are used with our fork of HuggingFace. At the time where this doc was created, the 2.1 releases were not officially out yet, and therefore, pre-releases links are used.
+```
+gcloud alpha compute tpus tpu-vm ssh ${TPU_NAME} \
```
Comment (jonb377) [inline SPMD_USER_GUIDE.md:40]: Just a note - I recently saw we have `queued-resources ssh`, which we can use for consistency.

File: SPMD_USER_GUIDE.md [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+### Cloud TPU VM Creation
+```
+export TPU_NAME=your-tpu-name
+export ZONE=us-central2-b
+export ACCELERATOR_TYPE=v4-32
+export RUNTIME_VERSION=tpu-ubuntu2204-base
+export PROJECT=your-project
+
+gcloud alpha compute tpus queued-resources create $TPU_NAME
+  \ --node-id=$TPU_NAME
```
Comment (jonb377) [inline SPMD_USER_GUIDE.md]: The `\` should be on the end of each line rather than beginning, right?

File: SPMD_USER_GUIDE.md [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+git clone -b llama2-google-next-training https://github.com/pytorch-tpu/transformers.git
+cd transformers
+sudo pip3 install -e . --user
+pip3 install datasets accelerate evaluate scikit-learn
+'
+```
+
+The last step for HF setup is to copy your Llama 2 config into the TPU VM.
+```
+gcloud compute tpus tpu-vm scp llama_2_7B.config $TPU_NAME:~/config.json --worker all --project $PROJECT --zone=$ZONE
```
Comment (jonb377) [inline SPMD_USER_GUIDE.md:58]: Same here - we can use `queued-resources scp`. No need to address this now, just a note.

File: examples/pytorch/language-modeling/run_clm.py [hunk 4]
```diff
@@ -308,6 +366,10 @@ def main():
     # Set seed before initializing model.
     set_seed(training_args.seed)
 
+    server = xp.start_server(9012)
+    logger.info('Profiling server started: {str(server)}')
```
Comment (jonb377) [inline examples/pytorch/language-modeling/run_clm.py]: My bad, but a small bug here - this should be an f-string: `f'Profiling.. {str(server)}'`

File: src/transformers/models/llama/modeling_llama.py [hunk 5]
```diff
@@ -1015,7 +1111,16 @@ def forward(
                 )
 
         # embed positions
+        # Is this the input to the model?
```
Comment (jonb377) [inline src/transformers/models/llama/modeling_llama.py]: nit - we can remove this

**Review by jonb377 — APPROVED** [review]
LGTM, thanks Jiewen! A couple of low-priority observations when reading through, no need to address now.

File: SPMD_USER_GUIDE.md (hunk 1 above)
Comment (alanwaketan) [reply SPMD_USER_GUIDE.md:40]: I guess it's probably doesn't matter...

File: SPMD_USER_GUIDE.md (hunk 2 above)
Comment (alanwaketan) [reply SPMD_USER_GUIDE.md]: Oops...

**Comment by alanwaketan** [comment]
Thanks @jonb377 for the review. Addressed all the comments.

**Comment by alanwaketan** [comment]
Close as the llama2-google-next-training is updated.
